issue-814-migrate-aws-privatelinks (#1392)
aiven/devportal: docs/platform/howto/use-aws-privatelinks.rst

Using AWS PrivateLink with Aiven services
=========================================

AWS `PrivateLink <https://aws.amazon.com/privatelink/>`__ brings Aiven
services to the selected virtual private cloud (VPC) in your AWS
account. In a traditional setup that uses `VPC
peering <https://help.aiven.io/en/articles/778836-using-vpc-peering>`__
, traffic is routed through an AWS VPC peering connection to your Aiven
services. With PrivateLink, you can create a VPC endpoint to your own
VPC and access an Aiven service from that. The VPC endpoint creates

> network interfaces (NIC) to the sub-nets and availability zones that you

Reviewer: ```suggestion
network interfaces (NIC) to the subnets and availability zones that you
```

choose and receives the private IP addresses that belong to the IP range
of your VPC. The VPC endpoint is routed to your Aiven service located in
one of Aiven's AWS accounts.

> You can enable PrivateLink for Aiven services located in project VPCs.

Reviewer: ```suggestion
You can enable PrivateLink for Aiven services located in project VPC.
```

Before you can set up AWS PrivateLink, `create a
VPC <https://help.aiven.io/en/articles/778836>`__ and launch the
services that you want to connect to that VPC. As there is no network

> routing between the VPCs, you can use any private IP range for the VPC,

Reviewer: ```suggestion
routing between the VPC, you can use any private IP range for the VPC,
```

unless you also want to connect to the project VPC using VPC peering
connections. This means that overlaps in the IP range are not an issue.

You can use either the `Aiven web console <https://console.aiven.io>`__
or the `Aiven CLI <https://github.com/aiven/aiven-client>`__ to set up

> AWS PrivateLink. You also need the AWS CLI to create a VPC endpoint.

Reviewer: Is AWS CLI the only way to create a VPC endpoint? Can they use [VPC Wizard](https://docs.aws.amazon.com/directoryservice/latest/admin-guide/gsg_create_vpc.html) from AWS console?

Author: The VPC wizard would be helpful to create the VPC, this point is referring to the VPC endpoint which is the interface created on the customer subnet to get connected to the Private link, yes it can be created from console: by following similar step listed in [Connect to an endpoint service as the service consumer 
](https://docs.aws.amazon.com/vpc/latest/privatelink/create-endpoint-service.html#share-endpoint-service ) 
I think we can mentioned that this could be done in the console, though I am not sure if sharing the link would cause some confusion.

> **Note:** Aiven for Apache Cassandra® and Aiven for M3 services do not

Reviewer: Since this might be an old Help article, could we pls verify that this **note** is still valid?

Author: I have verified that the note is still valid, I created Cassandra and M3 serviced in a project AWS VPC and the option to create private links is not available.

currently support AWS PrivateLink.

> #. | Create an AWS PrivateLink resource on the Aiven service.

Reviewer: ```suggestion
#. | Create an AWS PrivateLink resource on the Aiven service:
```

>    | The Amazon Resource Name (ARN) for the principals that are allowed

Reviewer: ```suggestion
   
   | The Amazon Resource Name (ARN) for the principals that are allowed
```

     the access scope for an entire AWS account, a given user account,
     or a given role. Only give permissions to roles that you trust, as
     an allowed role can connect from any VPC.

   -  Using the Aiven CLI, run the following command including your AWS
      account ID, the access scope, and the name of your Aiven service:

      ::

         $ avn service privatelink aws create --principal arn:aws:iam::$AWS_account_ID:$access_scope $Aiven_service_name

      For example:

      ::

         $ avn service privatelink aws create --principal arn:aws:iam::012345678901:user/mwf my-kafka

   -  Using the Aiven web console:

      #. Log in to the Aiven web console and select the service that you
         want to use.

      #. Select the **Network** tab and click **Create Privatelink** .

      #. | Enter the Amazon Resource Names (ARN) for the principals that
           you want to use, then click **Create** .

   | This creates an AWS network load balancer dedicated to your Aiven
     service and attaches it to an AWS VPC endpoint service that you can
     later use to connect to your account's VPC endpoint.

>    | The PrivateLink resource stays in the initial ``creating`` state

Reviewer: ```suggestion
   
   | The PrivateLink resource stays in the initial ``creating`` state
```

Author: implemented in latest commit

     for up to a few minutes while the load balancer is being launched.
     After the load balancer and VPC endpoint service have been created,
     the state changes to ``active`` and the ``aws_service_id`` and
     ``aws_service_name`` values are set.

#. In the AWS CLI, run the following command to create a VPC endpoint:

   ::

      $ aws ec2 --region eu-west-1 create-vpc-endpoint --vpc-endpoint-type Interface --vpc-id $your_vpc_id --subnet-ids $space_separated_list_of_subnet_ids --security-group-ids $security_group_ids --service-name com.amazonaws.vpce.eu-west-1.vpce-svc-0b16e88f3b706aaf1

   | 
   | Replace the ``--service-name`` value with the value shown next to
     **Network** > **AWS service name** in the Aiven web console or by
     running the following command in the Aiven CLI:

   ::

      $ avn service privatelink aws get aws_service_name

   | 
   | Note that for fault tolerance, you should specify a subnet ID for
     each availability zone in the region. The security groups determine
     the instances that are allowed to connect to the endpoint network

>      interfaces created by AWS into the specified sub-nets.

Reviewer: ```suggestion
     interfaces created by AWS into the specified subnets.
```

>    | Alternatively, you can create the VPC endpoint in the AWS web

Reviewer: ```suggestion
   
   | Alternatively, you can create the VPC endpoint in the AWS web
```

Author: implemented in latest commit

     console under **VPC** > **Endpoints** > **Create endpoint** . See
     the `AWS
     documentation <https://docs.aws.amazon.com/vpc/latest/userguide/vpce-interface.html#create-interface-endpoint>`__
     for details.
   | **Note:** For Aiven for Apache Kafka® services, the security group
     for the VPC endpoint must allow ingress in the port range
     ``10000-31000`` to accommodate the pool of Kafka broker ports used
     in our PrivateLink implementation.

>    | It takes a while before the endpoint is ready to use as AWS

Reviewer: ```suggestion
   
   | It takes a while before the endpoint is ready to use as AWS
```

>      provisions network interfaces to each of the sub-nets and connects

Reviewer: ```suggestion
     provisions network interfaces to each of the subnets and connects
```

Author: Vale does not recognize 'subnets' as a valid word, that is why I use sub-nets on the first place. What  is the best solution reword as sub-nets or use literal markup ``subnets``

     them to the Aiven VPC endpoint service. Once the AWS endpoint state
     changes to ``available`` , the connection is visible in Aiven.

> #. | Enable PrivateLink access for Aiven service components.

Reviewer: ```suggestion
#. | Enable PrivateLink access for Aiven service components:
```

>    | You can control each service component separately - for example,

Reviewer: ```suggestion
   
   | You can control each service component separately - for example,
```

      for the components that you want to enable. For example:

      ::

         $ avn service update -c privatelink_access.kafka=true $Aiven_service_name
         $ avn service update -c privatelink_access.kafka_connect=true $Aiven_service_name
         $ avn service update -c privatelink_access.kafka_rest=true $Aiven_service_name
         $ avn service update -c privatelink_access.schema_registry=true $Aiven_service_name

   -  In the Aiven web console:

      #. Select the **Overview** tab and scroll down to *Advanced
         configuration* .

      #. Click **Add configuration** , select the component that you
         want and switch it on.

         .. image:: /images/platform/howto/use-aws-privatelink_image1.png

      #. | Click **Save advanced configuration** .

   It takes a couple of minutes before connectivity is available after
   you enable a service component. This is because AWS requires an AWS
   load balancer behind each VPC endpoint service, and the target rules
   on the load balancer for the service nodes need at least two
   successful heartbeats before they transition from the ``initial``

>    state to ``healthy`` and are included in the load balancer's active

Reviewer: ```suggestion
   state to ``healthy`` and are included in the active forwarding rules of the load balancer.
```

>    forwarding rules.

Reviewer: ```suggestion

```
